packages/salesforcedx-vscode-core/src/commands/forceSourceDiff.ts

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2019, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import {
+  CliCommandExecutor,
+  Command,
+  DiffResultParser,
+  SfdxCommandBuilder
+} from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
+import {
+  CancelResponse,
+  ContinueResponse,
+  ParametersGatherer
+} from '@salesforce/salesforcedx-utils-vscode/out/src/types';
+import * as vscode from 'vscode';
+import { channelService } from '../channels';
+import { nls } from '../messages';
+import { notificationService, ProgressNotification } from '../notifications';
+import { taskViewService } from '../statuses';
+import { telemetryService } from '../telemetry';
+import { getRootWorkspacePath, hasRootWorkspace, OrgAuthInfo } from '../util';
+import {
+  SfdxCommandlet,
+  SfdxCommandletExecutor,
+  SfdxWorkspaceChecker
+} from './commands';
+
+export class ForceSourceDiffExecutor extends SfdxCommandletExecutor<{
+  filePath: string;
+}> {
+  public build(data: { filePath: string }): Command {
+    const commandBuilder = new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_source_diff_text'))
+      .withArg('force:source:diff')
+      .withLogName('force_source_diff')
+      .withFlag('--sourcepath', data.filePath)
+      .withJson();
+    return commandBuilder.build();
+  }
+
+  public async execute(
+    response: ContinueResponse<{ filePath: string }>
+  ): Promise<void> {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+
+    channelService.streamCommandStartStop(execution);
+    channelService.showChannelOutput();
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(realData => {
+      stdOut += realData.toString();
+    });
+
+    execution.processExitSubject.subscribe(async () => {
+      this.logMetric(execution.command.logName, startTime);
+
+      try {
+        const diffParser = new DiffResultParser(stdOut);
+        const diffParserSuccess = diffParser.getSuccessResponse();
+        const diffParserError = diffParser.getErrorResponse();
+
+        if (diffParser.isSuccessful() && diffParserSuccess) {
```

Comment:
do we need to check whether it's successful and whether diffParserSuccess is defined?

Reply:
Yes because of how I've implemented the parser methods where `diffParserSuccess` can be `undefined`. I'm open to changing the parser if this seems like over doing checks.

Reply:
It could probably just be `if (diffParserSuccess)` since `isSuccessful()` isn't used elsewhere